src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php

```diff
@@ -62,54 +62,74 @@ public function validate(mixed $entity, Constraint $constraint)
+        if (!\is_object($value)) {
+            throw new UnexpectedValueException($value, 'object');
         }
 
+        $entityClass = $value::class;
+
         if ($constraint->em) {
             $em = $this->registry->getManager($constraint->em);
 
             if (!$em) {
                 throw new ConstraintDefinitionException(sprintf('Object manager "%s" does not exist.', $constraint->em));
             }
         } else {
-            $em = $this->registry->getManagerForClass($entity::class);
+            $em = $this->registry->getManagerForClass($constraint->entityClass ?? $entityClass);
 
             if (!$em) {
-                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($entity)));
+                throw new ConstraintDefinitionException(sprintf('Unable to find the object manager associated with an entity of class "%s".', get_debug_type($value)));
             }
         }
 
-        $class = $em->getClassMetadata($entity::class);
+        try {
+            $repository = $em->getRepository($entityClass);
+            $isEntity = true;
+        } catch (ORMMappingException|PersistenceMappingException) {
+            $isEntity = false;
+        }
 
-        $criteria = [];
-        $hasIgnorableNullValue = false;
+        if (null !== $constraint->entityClass) {
+            /* Retrieve repository from given entity name.
+             * We ensure the retrieved repository can handle the entity
+             * by checking the entity is the same, or subclass of the supported entity.
+             */
+            $repository = $em->getRepository($constraint->entityClass);
+            $supportedClass = $repository->getClassName();
 
-        foreach ($fields as $fieldName) {
-            if (!$class->hasField($fieldName) && !$class->hasAssociation($fieldName)) {
-                throw new ConstraintDefinitionException(sprintf('The field "%s" is not mapped by Doctrine, so it cannot be validated for uniqueness.', $fieldName));
+            if ($isEntity && !$value instanceof $supportedClass) {
+                $class = $em->getClassMetadata($entityClass);
+                throw new ConstraintDefinitionException(sprintf('The "%s" entity repository does not support the "%s" entity. The entity should be an instance of or extend "%s".', $constraint->entityClass, $class->getName(), $supportedClass));
             }
+            $entityClass = $constraint->entityClass;
+        }
 
-            $fieldValue = $class->reflFields[$fieldName]->getValue($entity);
+        $class = $em->getClassMetadata($entityClass);
 
-            if (null === $fieldValue && $this->ignoreNullForField($constraint, $fieldName)) {
+        $criteria = [];
+        $hasIgnorableNullValue = false;
+
+        $fieldValues = $this->getFieldValues($value, $class, $fields, $isEntity);
+
+        foreach ($fieldValues as $entityFieldName => $fieldValue) {
```

Comment:
We could lessen the diff and make this easier to review (without sacrificing anything) by going back to the old loop `foreach ($fields as $fieldName) {` - then calling `$this->getFieldValue()` (singular) from inside the loop or inlining the logic inside the loop and not having the `getFieldValue()` method.

Reply:
Fields may now be a key=>value array (Example 2 from the description of PR). I moved the validation of the entity fields to the `getClassMetadata`. This method is also used to validate `$constraint->identifierFieldNames`. That way I eliminated some repeating code. Changed back `$entityFieldName` name to `$fieldName`; This lessen the diff.